### PR TITLE
ENH: special: translate `sph_bessel` to C++, refactor `cyl_bessel`

### PR DESCRIPTION
**Pull request by izaid**
This PR translates and refactors various Bessel functions, both spherical and cylindrical. Currently a work in progress.

### CODE REVIEW

**Comment by izaid** [comment]
@steppi This is done, ready for review.

**Review by steppi — COMMENTED** [review]
OK, there's a lot going on here @izaid . I'm going to give you a pass, because I know this is the last week you'll be able to devote to `SciPy` for a while, so I think it's fine to err on the side of getting as much as possible done. This was tough to review though due to a combination of many unrelated changes, and difficulty determining what was essential. The only change I've seen that is needed is ensuring `Python.h` is the first included header wherever it appears.

Let me know if the following list of changes is exhaustive.

1. A large number of formatting changes.
2. Passing references instead of pointers for ufuncs with multiple return values.
3. _spherical_bessel.pxd translated into C++.
4. Cylindrical bessel functions moved out of `amos` and into `bessel.h`.
     Can you give an account of what other changes, if any, were made besides moving
     these to a new file and changing some function names?
5. sf_error.cc and sf_error.c combined to a single file, sf_error.c.
6. New header `scipy/special/special.h` for functions where SciPy behavior deviate from behavior in C++ library (returning multiple values, where is from an independent function, runtime warning for noninteger parameters, returning `NAN` when result should be 0.
7. New header for exponential integrals, translation of scaled_exp1 into C++ header.

Did I miss anything?

File: scipy/special/ufunc.h
[the remark this answers is not in the document]
Comment (izaid) [reply scipy/special/ufunc.h:1]: Corrected!

File: scipy/special/special_wrappers.h [hunk 1]
```diff
@@ -24,30 +24,31 @@ npy_cdouble chyp2f1_wrap(double a, double b, double c, npy_cdouble z);
...
+npy_cdouble special_cexp1(npy_cdouble z);
+
+double special_expi(double x);
+npy_cdouble special_cexpi(npy_cdouble z);
 
 double struve_wrap(double v, double x);
-double itstruve0_wrap(double x);
-double it2struve0_wrap(double x);
+double special_itstruve0(double x);
+double special_it2struve0(double x);
```
[the remark this answers is not in the document]
Comment (izaid) [reply scipy/special/special_wrappers.h:37]: Yes, so the idea here is that the C wrappers are X_whatever for a C++ function X::whatever. Right now, X is "special", but we expect that to change in the future. This is pretty standard practice.

**Comment by izaid** [comment]
> OK, there's a lot going on here @izaid . I'm going to give you a pass, because I know this is the last week you'll be able to devote to `SciPy` for a while, so I think it's fine to err on the side of getting as much as possible done. This was tough to review though due to a combination of many unrelated changes, and difficulty determining what was essential. The only change I've seen that is needed is ensuring `Python.h` is the first included header wherever it appears.
> 
> Let me know if the following list of changes is exhaustive.
> 
>     1. A large number of formatting changes.
> 
>     2. Passing references instead of pointers for ufuncs with multiple return values.
> 
>     3. _spherical_bessel.pxd translated into C++.
> 
>     4. Cylindrical bessel functions moved out of `amos` and into `bessel.h`.
>        Can you give an account of what other changes, if any, were made besides moving
>        these to a new file and changing some function names?
> 
>     5. sf_error.cc and sf_error.c combined to a single file, sf_error.c.
> 
>     6. New header `scipy/special/special.h` for functions where SciPy behavior deviate from behavior in C++ library (returning multiple values, where is from an independent function, runtime warning for noninteger parameters, returning `NAN` when result should be 0.
> 
>     7. New header for exponential integrals, translation of scaled_exp1 into C++ header.
> 
> 
> Did I miss anything?



> OK, there's a lot going on here @izaid . I'm going to give you a pass, because I know this is the last week you'll be able to devote to `SciPy` for a while, so I think it's fine to err on the side of getting as much as possible done. This was tough to review though due to a combination of many unrelated changes, and difficulty determining what was essential. The only change I've seen that is needed is ensuring `Python.h` is the first included header wherever it appears.
> 
> Let me know if the following list of changes is exhaustive.
> 
>     1. A large number of formatting changes.
> 
>     2. Passing references instead of pointers for ufuncs with multiple return values.
> 
>     3. _spherical_bessel.pxd translated into C++.
> 
>     4. Cylindrical bessel functions moved out of `amos` and into `bessel.h`.
>        Can you give an account of what other changes, if any, were made besides moving
>        these to a new file and changing some function names?
> 
>     5. sf_error.cc and sf_error.c combined to a single file, sf_error.c.
> 
>     6. New header `scipy/special/special.h` for functions where SciPy behavior deviate from behavior in C++ library (returning multiple values, where is from an independent function, runtime warning for noninteger parameters, returning `NAN` when result should be 0.
> 
>     7. New header for exponential integrals, translation of scaled_exp1 into C++ header.
> 
> 
> Did I miss anything?

That's pretty comprehensive! A lot of refactoring and things to make life easier, plus translation of the spherical Bessel functions. I've made the changes requested now too.

File: scipy/special/ufunc.h
Comment (steppi) [inline scipy/special/ufunc.h:1]: It looks like Python.h still needs to be moved to be the first include here.

Comment (izaid) [reply scipy/special/ufunc.h:1]: Done. Why does it need to be before headers from the C / C++ library though? Is this just a SciPy convention?

Comment (steppi) [reply scipy/special/ufunc.h:1]: It causes errors downstream in some cases. It’s a Python C API convention.

**Review by steppi — APPROVED** [review]
I'm happy with this now.

**Pull request merged**